Add RemoteMongoClient

RedBeard0531 opened the pull request:
<!-- Make sure to assign one and only one Type (`T:`) label.
 Select reviewers if ready for review. Our bot will automatically assign you. -->

## What, How & Why?
<!-- Describe the changes and give some hints to guide your reviewers if possible. -->
<!-- E.g. reference to other repos: This closes realm/realm-sync#??? -->


## ☑️ ToDos
<!-- Add your own todos here -->
* [X] 📝 Changelog entry
* ~[ ] 📝 `Compatibility` label is updated or copied from previous entry~
* [X] 🚦 Tests
* ~[ ] 📝 Public documentation PR created or is not necessary~
* ~[ ] 💥 `Breaking` label has been applied or is not necessary~

*If this PR adds or changes public API's:*
* ~[ ] typescript definitions file is updated~ 
* [x] jsdoc files updated
* ~[ ] Chrome debug API is updated if API is available on React Native~

RedBeard0531 commented on lib/mongo_client.js:
```diff
@@ -0,0 +1,211 @@
+////////////////////////////////////////////////////////////////////////////
```
This is a JS-ified copy of https://github.com/realm/realm-js/blob/v10/packages/realm-web/src/services/RemoteMongoDBService.ts with only the RemoteMongoDBCollection type and without the factory functions. The constructor was lightly modified to accept the `functions` object directly rather than constructing it itself.

RedBeard0531 commented on lib/user.js:
```diff
@@ -31,9 +33,23 @@ const instanceMethods = {
     },
 
-    callFunction(name, args) {
+    callFunction(name, args, service = undefined) {
```
Added the optional `service` param to support `remoteMongoClient()`. Also now using `cleanArgs()` preprocessing similar taken from https://github.com/realm/realm-js/blob/11f690217253a095cf605779d006f0be9f7cb6a1/packages/realm-web/src/FunctionsFactory.ts#L51-L68

RedBeard0531 commented on lib/user.js:
```diff
@@ -55,17 +71,41 @@ const instanceMethods = {
+        };
+    },
+
+    _functionsOnService(service) {
```
This is basically the same as the old `get functions()` getter but now takes a `service` and uses the property name filtering logic taken from https://github.com/realm/realm-js/blob/11f690217253a095cf605779d006f0be9f7cb6a1/packages/realm-web/src/FunctionsFactory.ts#L149-L158

blagoev commented on lib/user.js:
```diff
@@ -31,9 +33,23 @@ const instanceMethods = {
+
         return new Promise((resolve, reject) => {
-            this._callFunction(name, args, (result, error) => {
+            this._callFunction(name, args, service, (result, error) => {
```
this does not resolve

blagoev replied:
ah.. ignore. the github ui have failed me

RedBeard0531 replied:
I have a patch locally that makes this just `return promisify(cb => this._callFunction(name, args, service, cb));` so you won't have to worry about that in the future :)

blagoev commented on lib/user.js:
```diff
@@ -55,17 +71,41 @@ const instanceMethods = {
-                return (...args) => {
-                    return user.callFunction(name, args);
+        return {
+            db(dbName) {
```
Is this calling a function or returning an object?

RedBeard0531 replied:
`remoteMongoClient()` is returning an object with a single method `db()`, which returns and object with the single method `collection()` which returns a RemoteMongoDBCollection object.

It may be clearer if you see a usage example: https://github.com/realm/realm-js/blob/d13d14eb38c7b27ad7b3c5fd2ceb6d479d2f704e/tests/js/user-tests.js#L158-L165

blagoev replied:
got it

blagoev commented on src/js_user.hpp:
```diff
@@ -242,18 +242,23 @@ void UserClass<T>::link_credentials(ContextType ctx, ObjectType this_object, Arg
 template<typename T>
 void UserClass<T>::call_function(ContextType ctx, ObjectType this_object, Arguments& args, ReturnValue &) {
-    args.validate_count(3);
+    args.validate_count(4);
```
are all arguments required. Isn't something optional

RedBeard0531 replied:
Optional in the user-facing JS API. Not optional in the internal C++ API. Well semi-optional since it accepts an explicit `undefined`.

blagoev replied:
👍

blagoev commented on src/js_user.hpp:
```diff
@@ -242,18 +242,23 @@ void UserClass<T>::link_credentials(ContextType ctx, ObjectType this_object, Arg
     auto name = Value::validated_to_string(ctx, args[0], "name");
     auto call_args_js = Value::validated_to_array(ctx, args[1], "args");
-    auto callback = Value::validated_to_function(ctx, args[2], "callback");
+    auto service = Value::is_undefined(ctx, args[2])
```
if service is optional isn't it better to be last parameter. Or this is just because to follow JS convention where callback is last

RedBeard0531 replied:
I generally try to follow the callback-last API principle since it makes calling code clearer. And in this case the only caller always explicitly provides all arguments, one of them is just frequently the value `undefined`.

blagoev replied:
👍

kraenhansen commented on tests/js/user-tests.js:
```diff
@@ -150,6 +150,21 @@ module.exports = {
   },
 
+  async testRemoteMongoClient() {
+    let app = new Realm.App(appConfig);
```
Not really testing much, but I guess that's fine for now, we should reuse the Realm Web integration tests long term.

RedBeard0531 replied:
Yeah, the tests were just to ensure that everything was hooked up right and working sanely. Agree that we should one have one comprehensive test long term.

kraenhansen commented on tests/js/user-tests.js:
```diff
@@ -150,6 +150,21 @@ module.exports = {
+    let collection = mongo.db('test').collection('testRemoteMongoClient');
+
+    await collection.deleteMany({});
+    await collection.insertOne({hello: "world"});
```
Is this relying on a cleanup defined elsewhere or does this require a manual reset of the database?

RedBeard0531 replied:
Line 161 does cleanup so the rest of the test always runs in a clean state.

kraenhansen replied:
🤦 Never mind ...

RedBeard0531 commented:
I added jsdoc documentation for this, mostly copied and modified from the .d.ts file. Let me know if you are happy with the result. Or if you would prefer me to open a separate PR for that rather than adding it to this one.

bmunkholm commented on docs/sync.js:
```diff
@@ -762,3 +776,208 @@ class Worker {
+    /**
+     * Finds a document which matches the provided query and performs the desired update to individual fields.
+     *
+     * @param {object} filter An filter applied to narrow down the results.
```
```suggestion
     * @param {object} filter A filter applied to narrow down the results.
```

bmunkholm commented on docs/sync.js:
```diff
@@ -762,3 +776,208 @@ class Worker {
+    /**
+     * Finds a document which matches the provided filter and replaces it with a new document.
+     *
+     * @param {object} filter An filter applied to narrow down the results.
```
```suggestion
     * @param {object} filter A filter applied to narrow down the results.
```

bmunkholm commented on docs/sync.js:
```diff
@@ -762,3 +776,208 @@ class Worker {
+    /**
+     * Finds a document which matches the provided filter and deletes it
+     *
+     * @param {object} filter An filter applied to narrow down the results.
```
```suggestion
     * @param {object} filter A filter applied to narrow down the results.
```

bmunkholm commented on docs/sync.js:
```diff
@@ -762,3 +776,208 @@ class Worker {
+    /**
+     * Updates a single document matching the provided filter in this collection.
+     *
+     * @param {object} filter An filter applied to narrow down the results.
```
```suggestion
     * @param {object} filter A filter applied to narrow down the results.
```

bmunkholm commented on docs/sync.js:
```diff
@@ -762,3 +776,208 @@ class Worker {
+    /**
+     * Updates multiple documents matching the provided filter in this collection.
+     *
+     * @param {object} filter An filter applied to narrow down the results.
```
```suggestion
     * @param {object} filter A filter applied to narrow down the results.
```
